addons/survey/models/survey_survey.py

```diff
@@ -181,6 +185,42 @@ def _read_group_stage_ids(self, stages, domain, order):
         stage_ids = stages._search([], order=order, access_rights_uid=SUPERUSER_ID)
         return stages.browse(stage_ids)
 
+    @api.model
+    def create(self, vals):
+        if 'certification_badge_id' in vals:
+            survey = super(Survey, self).create(vals)
+            goal = self.env['gamification.goal.definition'].create({
+                'name': survey.title,
+                'description': "%s certification passed" % survey.title,
+                'domain': "['&', ('survey_id', '=', %s), ('quizz_passed', '=', True)]" % survey.id,
+                'computation_mode': 'count',
+                'display_mode': 'boolean',
+                'model_id': self.env.ref('survey.model_survey_user_input').id,
+                'condition': 'higher',
+                'batch_mode': True,
+                'batch_distinctive_field': self.env.ref('survey.field_survey_user_input__partner_id').id,
+                'batch_user_expression': 'user.partner_id.id'
+            })
+            if not survey.certification_badge_id.challenge_ids:
+                challenge = self.env['gamification.challenge'].create({
+                    'name': '%s challenge certificate' % survey.title,
+                    'reward_id': survey.certification_badge_id.id,
+                    'state': 'inprogress',
+                    'period': 'once',
+                    'category': 'certification',
+                    'reward_realtime': True,
+                    'report_message_frequency': 'never',
+                    'user_domain': [('karma', '>', 0)],
+                    'visibility_mode': 'personal'
+                })
+            self.env['gamification.challenge.line'].create({
+                'definition_id': goal.id,
+                'challenge_id': challenge.id,
```

Comment:
Pay attention that challenge does not necessarily exist in the current code; however as code has to change a bit just correctly fetch existing challenge or newly-created one.

Reply:
correct!